pkg/xen-tools/initrd/init-initrd

```diff
@@ -19,182 +20,263 @@ else
-mount -t tmpfs -o nodev,nosuid,size=20% tmp /mnt/rootfs/tmp
-mount -t mqueue -o nodev,nosuid,noexec none /mnt/rootfs/dev/mqueue
-ln -s /proc/self/fd /mnt/rootfs/dev/fd
-ln -s /proc/self/fd/0 /mnt/rootfs/dev/stdin
-ln -s /proc/self/fd/1 /mnt/rootfs/dev/stdout
-ln -s /proc/self/fd/2 /mnt/rootfs/dev/stderr
-
-ip=`cat /proc/cmdline | grep -o '\bip=[^ ]*' | cut -d = -f 2`
-gw=`cat /proc/cmdline | grep -o '\bgw=[^ ]*' | cut -d = -f 2`
-route=`cat /proc/cmdline | grep -o '\broute=[^ ]*' | cut -d = -f 2`
-pvcalls=`cat /proc/cmdline | grep  'pvcalls=1'`
-dhcp=`cat /proc/cmdline | grep  'dhcp=1'`
-# shellcheck disable=SC2002,SC2006
-chrony=`cat /proc/cmdline | grep 'chrony=1'`
-if test "$pvcalls"
-then
-        mkdir -p /mnt/rootfs/etc
-        echo "nameserver 8.8.8.8" > /mnt/rootfs/etc/resolv.conf
-        echo "options use-vc" >> /mnt/rootfs/etc/resolv.conf
-elif test "$ip"
-then
-    # ipv4
-    if [ "$ip" != "${ip#*[0-9].[0-9]}" ]
+
+    mkdir "$MNT" >/dev/null 2>&1
+    # XXX where are the other mounts on 9p?
+    if [ "$root" = "9p" ]; then
+        mount -t 9p -o msize=131072,trans="$HYPER_BUS",version=9p2000.L,cache=mmap share_dir "$MNT"
+        ret=$?
+    else
+        mount "$root" "$MNT"
+        ret=$?
+    fi
+    if [ $ret != 0 ] ; then
+        if [ "$oci" = 0 ]; then
+            echo "Failed to mount root"
+            # Should we go to the wait at the end
+            exit 1
+        else
+            echo "Failed to mount on $MNT"
+            continue
+        fi
+    fi
+    echo "Mounted $MNT"
+
+    mkdir "$MNT"/rootfs/dev >/dev/null 2>&1
+    mkdir -p "$MNT"/rootfs/sys/fs/cgroup >/dev/null 2>&1
+    mkdir "$MNT"/rootfs/proc >/dev/null 2>&1
+    mkdir /dev/pts >/dev/null 2>&1
+    mkdir /dev/shm >/dev/null 2>&1
+    mkdir /dev/mqueue >/dev/null 2>&1
+    if [ -d "$MNT"/rootfs/dev/eve ]; then
+        mkdir -p /dev/eve
+        mount -o bind "$MNT"/rootfs/dev/eve /dev/eve
+    fi
+    mount -o rbind /dev "$MNT"/rootfs/dev
+    mount -o bind /sys "$MNT"/rootfs/sys
+    mount -t tmpfs -o uid=0,gid=0,mode=0755 cgroup "$MNT"/rootfs/sys/fs/cgroup
+    awk '!/^#/ { if ($4 == 1) print $1 }' /proc/cgroups | while IFS= read -r sys
+    do
+        cgroup="$MNT/rootfs/sys/fs/cgroup/$sys"
+        mkdir -p "$cgroup"
+        if ! mountpoint -q "$cgroup"; then
+            if ! mount -n -t cgroup -o "$sys" cgroup "$cgroup"; then
+                rmdir "$cgroup" || true
+            fi
+        fi
+    done
+    if [ -e "$MNT"/rootfs/sys/fs/cgroup/memory/memory.use_hierarchy ]; then
+        echo 1 > "$MNT"/rootfs/sys/fs/cgroup/memory/memory.use_hierarchy
+    fi
+    mount -t devpts -o gid=5,mode=0620,noexec,nosuid devpts "$MNT"/rootfs/dev/pts
+    mount -t tmpfs -o nodev,nosuid,noexec,size=20% shm "$MNT"/rootfs/dev/shm
+    mount -t tmpfs -o nodev,nosuid,size=20% tmp "$MNT"/rootfs/tmp
+    mount -t mqueue -o nodev,nosuid,noexec none "$MNT"/rootfs/dev/mqueue
+    ln -s /proc/self/fd "$MNT"/rootfs/dev/fd
```

Comment:
Sharing all these descriptors with all containers wouldn't mess things up? Without a mux we will have mixed outputs on stdout, which might not be critical for now, but what about the stdin? Do we care about it?

Reply:
The use case for this experiment is when some sharing is ok, but I'll look at the list and see what makes sense to separate.
If the direction in this PR is useful (with its limitations) it might be a useful stepping stone to running a collection of containers (a pod) in a VM using something existing like kata or k3s.